### PR TITLE
Improve decimal formatter functions

### PR DESCRIPTION
**Pull request by jinchung**
## What changed (plus any additional context for devs)
1. The `valueBasedDecimalFormatter` now takes into account the magnitude of the asset amount by default. It balances between significant decimals, value-based decimals, and magnitude.
2. Moved out the `valueBasedDecimalFormatter` into a separate file to make it easier to test
3. Added test cases for the `valueBasedDecimalFormatter` for various asset amounts and asset prices, no prices, stablecoins, etc
4. Added an optional param to `valueBasedDecimalFormatter` that the `niceIncrementFormatter` can pass in such that it guarantees that the `valueBasedDecimalFormatter` is always at least +1 more than the `niceIncrementFormatter` number of decimals. This helps resolve scenarios where the "jumps" between increments might end up rounding funnily on the snapping points.
5. Removed `incrementDecimalPlaces` and `niceIncrement` params to the `niceIncrementFormatter` since we can calculate this within the formatter function itself and don't use it elsewhere.

## What to test
- updating input amount, slider, output amount behaves as expected (just with nicer formatting)
- typing in "0.0000..." and trying to type "0000" behaves the same (where "0000" just stays as "0" and "0.0000" is left as is)
- for assets with super small balances, there shouldn't be any weird jumps in number when going from 49% -> 50% -> 51% (which was possible before)

## Screen recordings / screenshots
Example of weird jumps in value using slider on assets with super small balances BEFORE this PR:
https://github.com/rainbow-me/rainbow/assets/1285228/35393188-3552-4a68-867e-3853bdc69dd9

Example of smooth values using slider on assets with super small balances BEFORE this PR:
https://github.com/rainbow-me/rainbow/assets/1285228/8a891544-e92f-4487-ad41-c999ecd4d7b5

Example of formatting BEFORE this PR:
IMG_7431
IMG_7429
IMG_7430

Example of formatting AFTER this PR:
eth=25
eth-50
eth-75

### CODE REVIEW

File: src/__swaps__/utils/swaps.ts [hunk 1]
```diff
@@ -233,139 +232,53 @@ export function trimTrailingZeros(value: string) {
...
   'worklet';
+  const niceIncrement = findNiceIncrement(inputAssetBalance);
+  const incrementDecimalPlaces = countDecimalPlaces(niceIncrement);
 
   if (percentageToSwap === 0 || equalWorklet(niceIncrement, 0)) return '0';
   if (percentageToSwap === 0.25) {
     const amount = mulWorklet(inputAssetBalance, 0.25);
     return valueBasedDecimalFormatter({
       nativePrice: inputAssetNativePrice,
+      niceIncrementMinimumDecimals: incrementDecimalPlaces,
```
Comment (jinchung) [inline src/__swaps__/utils/swaps.ts:259]: This is point #4: Added an optional param to valueBasedDecimalFormatter that the niceIncrementFormatter can pass in such that it guarantees that the valueBasedDecimalFormatter is always at least +1 more than the niceIncrementFormatter number of decimals. This helps resolve scenarios where the "jumps" between increments might end up rounding funnily on the snapping points.

File: src/__swaps__/utils/decimalFormatter.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      minimumDecimalPlaces = Math.max(stablecoinMin, significantDecimals);
+      maximumDecimalPlaces = Math.max(minBasedOnOrderOfMag, minDecimalsForOneCent, significantDecimals + 1);
+    } else {
+      minimumDecimalPlaces = stablecoinMin;
+      maximumDecimalPlaces = Math.max(minBasedOnOrderOfMag, minDecimalsForOneCent - orderOfMagnitude);
+    }
+
+    return {
+      minimumDecimalPlaces,
+      maximumDecimalPlaces: Math.max(
```
Comment (jinchung) [inline src/__swaps__/utils/decimalFormatter.ts:58]: This just ensures that after all of our ideal calculations, we ensure that:
1. incorporate any `precisionAdjustment` here if wanted
2. ensure that if there's a value passed in from the `niceIncrementFormatter` that we at least are +1 decimal above that
3. the max should always be greater than the min decimals calculated

File: src/__swaps__/utils/decimalFormatter.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    let maximumDecimalPlaces = MAXIMUM_SIGNIFICANT_DECIMALS;
+    const stablecoinMin = isStablecoin ? STABLECOIN_MINIMUM_SIGNIFICANT_DECIMALS : 0;
+
+    const minBasedOnOrderOfMag = orderOfMagnitude > 2 ? 0 : 2;
+    if (orderOfMagnitude < 1) {
+      minimumDecimalPlaces = Math.max(stablecoinMin, significantDecimals);
+      maximumDecimalPlaces = Math.max(minBasedOnOrderOfMag, minDecimalsForOneCent, significantDecimals + 1);
+    } else {
+      minimumDecimalPlaces = stablecoinMin;
+      maximumDecimalPlaces = Math.max(minBasedOnOrderOfMag, minDecimalsForOneCent - orderOfMagnitude);
```
Comment (jinchung) [inline src/__swaps__/utils/decimalFormatter.ts:53]: this balances the `orderOfMagnitude` with the `minDecimalsForOneCent`

File: src/__swaps__/safe-math/SafeMath.ts [hunk 4]
```diff
@@ -202,6 +202,32 @@ export function modWorklet(num1: string | number, num2: string | number): string
...
+  }
+
+  // Convert the number to a string to handle large numbers and fractional parts
+  const numStr = num.toString();
+
+  // Split the number into integer and fractional parts
+  const [_, fractionalPart] = numStr.split('.');
+
+  // Handle fractional parts
+  if (fractionalPart) {
```
Comment (jinchung) [inline src/__swaps__/safe-math/SafeMath.ts:219]: I created a new function here for now as the `orderOfMagnitudeWorklet` which is used in other places doesn't exactly capture what we sometimes care about for counting significant decimals. e.g, in this function, I'm curious what the first significant decimal is, regardless of what's to the left of the decimal.

File: src/__swaps__/safe-math/SafeMath.ts [hunk 5]
```diff
@@ -202,6 +202,32 @@ export function modWorklet(num1: string | number, num2: string | number): string
...
+    // Find the first non-zero digit in the fractional part
+    for (let i = 0; i < fractionalPart.length; i++) {
+      if (fractionalPart[i] !== '0') {
+        return i + 1;
+      }
+    }
```
Comment (greg-schrammel) [inline src/__swaps__/safe-math/SafeMath.ts:225]: I think `fractionalPart.indexOf('0')` works?

**Review by greg-schrammel — APPROVED** [review]

**Review by ibrahimtaveras00 — APPROVED** [review]
Looks good on both OS's 👍🏽

**Pull request merged**